woocommerce/woocommerce-android
# Domain change: Contact form and purchase checkout

Fixes #8407, a subtask of #8238.

First, I want to apologize for the massive PR 🙇. 3/4 of it is the Domain contact from and the associated code, that I copied over from WPAndroid. That's why it's still using XML layout. I wanted to save some time and reuse the existing code since it's been tested and it already contained the transaction logic to purchase a domain using free credits + also things like contact information pre-filling, form validation, etc. The coupling is fairly loose so we can eventually go back and replace contact form with Jetpack Compose, if we want to. 

The domain registration/purchase flow is now as follows:

Domain dashboard screen -> Domain search scree -> `if (hasFreeCredit)` Free domain registration screen (w/ domain contact form) `else` -> Domain purchase screen (webview) -> Purchase success screen

I didn't do much refactoring of the Free domain registration screen besides some minimum changes required to integrate it in the purchase flow. I've thought about splitting the PR somehow but it wouldn't make much difference if I wanted to keep the logical parts of the contact form together. 

If I can suggest a way to review this - ignore the `FreeDomainRegistrationFragment`, its layout XML, `FreeDomainRegistrationViewModel` and `DomainPhoneNumberUtils` at first and leave it for the last as that's the copied code. 

**To test:**

Domain registration using free credit
1. Login using an admin account
2. Select a store that's eligible for a free domain (or create a new one from the app)
3. Go to Settings -> Domains -> Claim a free domain
4. Find a non-premium domain that's free for the first year
5. Tap on Continue
6. If the logged-in WP user has an address stored, it'll be pre-filled
7. If you try to clear any required field like a name, phone number or a country, a validation error is displayed
8. Once the contact details are entered, tap on the Register button
9. Wait for a while until the registration completes
10. Notice the Success screen is shown with a new domain name 


Domain registration via WebView purchase
1. Login using an admin account (give yourself free credits first in SA)
2. Select a store that's _not_ eligible for a free domain 
3. Go to Settings -> Domains
4. Tap on Search for a domain button
5. Select a paid domain
6. Tap on Continue
7. Notice a webview is opened with a checkout website that shows a new domain in a shopping cart
8. Complete the contact information, if it's not filled out already
9. Complete the purchase 
10. Notice the Success screen is shown with a new domain name
